client.go

```diff
@@ -1,82 +1,235 @@
 	log := config.LoggerFactory.NewLogger("turnc")
-	network := "udp4"
 
 	if config.Net == nil {
 		config.Net = vnet.NewNet(nil) // defaults to native operation
 	} else {
 		log.Warn("vnet is enabled")
 	}
 
-	c := &Client{
-		net: config.Net,
-		log: log,
+	var stunServ, turnServ net.Addr
+	var err error
+	if len(config.STUNServerAddr) > 0 {
+		log.Debugf("resolving %s", config.STUNServerAddr)
+		stunServ, err = config.Net.ResolveUDPAddr("udp4", config.STUNServerAddr)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("stunServ: %s", stunServ.String())
+	}
+	if len(config.TURNServerAddr) > 0 {
+		log.Debugf("resolving %s", config.TURNServerAddr)
+		turnServ, err = config.Net.ResolveUDPAddr("udp4", config.TURNServerAddr)
+		if err != nil {
+			return nil, err
+		}
+		log.Debugf("turnServ: %s", stunServ.String())
 	}
 
-	var err error
-	c.conn, err = c.net.ListenPacket(network, config.ListeningAddress)
-	if err != nil {
-		return nil, errors.Wrap(err, fmt.Sprintf("failed to listen on %s", config.ListeningAddress))
+	c := &Client{
+		conn:     config.Conn,
+		stunServ: stunServ,
+		turnServ: turnServ,
+		username: stun.NewUsername(config.Username),
+		password: config.Password,
+		realm:    stun.NewRealm(config.Realm),
+		net:      config.Net,
+		trMap:    map[string]*transaction{},
+		rto:      defaultRTO,
+		log:      log,
 	}
 
 	return c, nil
 }
 
-// SendSTUNRequest sends a new STUN request to the serverIP with serverPort
-func (c *Client) SendSTUNRequest(serverIP net.IP, serverPort int) (net.Addr, error) {
-	c.mux.Lock()
-	defer c.mux.Unlock()
+// TURNServerAddr return the TURN server address
+func (c *Client) TURNServerAddr() net.Addr {
+	return c.turnServ
+}
 
-	c.log.Debug("sending STUN request")
-	if err := sendStunRequest(c.conn, serverIP, serverPort); err != nil {
-		return nil, err
+// STUNServerAddr return the STUN server address
+func (c *Client) STUNServerAddr() net.Addr {
+	return c.stunServ
+}
+
+// Username returns username
+func (c *Client) Username() stun.Username {
+	return c.username
+}
+
+// Realm return realm
+func (c *Client) Realm() stun.Realm {
+	return c.realm
+}
+
+// WriteTo sends data to the specified destination using the base socket.
+func (c *Client) WriteTo(data []byte, to net.Addr) (int, error) {
+	return c.conn.WriteTo(data, to)
+}
+
+// Listen will have this client start listening on the conn provided via the config.
+// This is optional. If not used, you will need to call HandleInbound method
+// to supply incoming data, instead.
+func (c *Client) Listen() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.conn == nil {
+		conn, err := c.net.ListenPacket("udp4", "0.0.0.0:0")
+		if err != nil {
+			return err
+		}
+		c.conn = conn
 	}
 
-	packet := make([]byte, 1500)
-	c.log.Debug("wait for STUN response...")
-	size, _, err := c.conn.ReadFrom(packet)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to read packet from udp socket")
+	if c.isListening {
```

Comment:
I would put `isListening` inside a function (and personally would prefer an atomic) then we don't have to worry about taking/releasing a mutex!

Reply:
Now, conn is provided from outside. It is turn.Client's responsibility to close the conn. (isListening is no longer needed)